uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Service/WorkPatternCalendar.php

```diff
@@ -36,8 +36,19 @@ class CRM_HRLeaveAndAbsences_Service_WorkPatternCalendar {
    */
   private $defaultWorkPattern;
 
-  public function __construct($contactID, AbsencePeriod $absencePeriod, JobContractService $jobContractService) {
+  /**
+   * CRM_HRLeaveAndAbsences_Service_WorkPatternCalendar constructor.
+   *
+   * @param int $contactID
+   * @param string $startDate
+   * @param string $endDate
+   * @param \CRM_HRLeaveAndAbsences_Service_JobContract $jobContractService
+   */
+  public function __construct($contactID, $startDate, $endDate, JobContractService $jobContractService) {
     $this->contactID = $contactID;
+    $absencePeriod = new AbsencePeriod();
+    $absencePeriod->start_date = CRM_Utils_Date::processDate($startDate, null, false, 'Y-m-d');
+    $absencePeriod->end_date = CRM_Utils_Date::processDate($endDate, null, false, 'Y-m-d');
```

Comment:
> as it also does does things already in the constructor which we don't need, like modifying the end date.
You do you mean by this? Are you talking about the `modify('+1 day')` statement? This is required to "fix" the unusual behavior of PHP's default DatePeriod class. It doesn't actually change the given end date.

I still think we should use the `BasicDatePeriod` class. As I mentioned, having an incomplete entity is a problem. A huge one (I know there are other parts of the code where we use this, but it's something that should be avoided when we have alternatives). 

There's also the problem of the meaning. The calendar now gives you dates for a period, which might be:
- A fraction of an Absence Period
- Exactly the same as an Absence Period
- MORE than an Absence Period (would this would work?)

But the thing is, it's not an Absence Period anymore, but just a generic period of dates.

So here is what I would do:

- Move the `adjustDatesToMatchPeriodDates` method to the `BasicDatePeriod` class. The method is generic enough that this extraction should be very simple (also, a good signal that it should not belong to the AbsencePeriod BAO). Also, I think that instead of returning an array, this method should return a new `BasicDatePeriod` instance
- To get easy access to the `start` and `end` dates, you can add getters for the `BasicDatePeriod` class.
- Change the `WorkCalendar` constructor to accept a `BasicDatePeriod` instance instead of these typeless `startDate` and `endDate` params